man/man1/cdf2bin.1
.TH cdf2bin 1  "USER COMMANDS"
.SH NAME
cdf2bin \- convert NetCDF files into raw binary data
.SH SYNOPSIS
.B cdf2bin
[\-h] [\-v] [\-swap] [\-units u] [\-var name] [\-map name] [\-clm] [\-byte] [\-cellsize size] [\-floatgrid] [\-doublegrid] [\-json] [\-o binfile]
gridfile netcdffile ...
.SH DESCRIPTION
Program converts NetCDF output data into raw binary output data.
.SH OPTIONS
.TP
\-h,\--help
Print a short help text.
.TP
\-v,\--verbose
Print name of NetCDF files.
.TP
\-swap
Byte order of grid file has to be swapped.
.TP
\-units u
Units to convert data from NetCDF file.
.TP
\-var name
Variable name in NetCDF file.
.TP
\-map name
Name of map in NetCDF file. If specified the map is read and put into the JSON metafile.
.TP
\-floatgrid
Set datatype of grid file to float, default is short.
.TP
\-doublegrid
Set datatype of grid file to double, default is short.
.TP
\-clm
Gridfile is in CLM format, default is raw.
.TP
\-byte
Output data is converted to byte.
.TP
\-cellsize size
Cell size in grid file, default is 0.5.
.TP
\-json
JSON metafile is created with suffix '.json'.
.TP
\-o binfile
Filename of raw binary data file written. Default is out.bin.
.TP
.I gridfile
Filename of binary grid data file consisting of short, float or double values or clm file if -clm option is set.
.TP
.I netcdffile
Filename of NetCDF file(s) converted.
.SH EXAMPLES
.TP
Create output file for vegetation carbon:
.B cdf2bin
-var vegc -o vegc1901-2010.bin grid.bin vegc.nc
.PP
.SH EXIT STATUS
.B cdf2bin
returns a zero exit status if the binary file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
bin2cdf(1), cmpbin(1), cdf2clm(1), cdf2grid(1), country2cdf(1), regridlpj(1), lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: David Hötten (davidho@pik-potsdam.de)

### Added

- Warning message added in `lpjml`, `lpjcheck`,`cdf2soil`, `txt2grid` and `cdf2coord` if cell size cannot be resolved by grid data in short format.
- Function `isfloatcoord()` added in `coord.c` to check whether coordinate must be stored as a float variable, test function `test_isfloatcoord.c` added.
- Test functions for `vmaxlimit_tree`, `vmaxlimit_grass`, `vmaxlimit_crop`, `bisect`, `leftmostzero` and `getbufferavg` functions added.

### Changed

- Error about first simulation year is less than first year of climate data replaced by warning in `initclimate.c`.
- First and last simulation year in spinup changed from 1901 to 1840 in `lpjml_config.cjson`.
- Number of spinup years changed from 3500 to 3439 in `lpjml_config.cjson` in order to get bit-identical results.
- Test files for unit test added to tar and zip file in `Makefile`.

### Removed

- Unused source file `src/lpj/part.c` removed.

### Fixed

- Expected number of bands set correctly for raw input files in `openinputdata.c`.
- Hard-coded filename `test.lpj` replaced by temporary filename in `test_append.c`, `test_getnoread.c`, `test_indexarray.c`, `test_isdefined.c`, `test_isnull.c`, `test_restart.c`, and `test_writename.c` in order to allow parallel execution of tests.
- Arguments corrected in calls of lpjml functions in `lpj_poem.c`.
- Missing `-print_noread` option added in `INSTALL`.
- `setupannual_grid.c` corrected to compile with `-DIMAGE -DCOUPLED` flag.
- Error message for different datatypes corrected in `lpcat.c`.
- Variables initialized to zero and NULL in `phenology_gsi.c`, `lpjcat.c` and `restart2yaml.c` to avoid warnings in `gcc`.
- Man page for `json2restart` corrected.
- Missing option `-nooutput` added to man page of `lpjfiles`.
- Help text for `-h` option of `printclm` corrected.
- Check for NULL pointer for `pfts` array corrected in `lpjcat.c`.
- Double entry of `getlag.$O` removed in `src/soil/Makefile`.


## [5.10.1] - 2025-09-30

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review: Maik Billing (billing@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- Variable `phen` added to output in `fprintpft.c`.
- Compile option `NRECOVERY_COST` added to `INSTALL` and man page of `configure.sh`.
- Check for valid `swc_bnf` parameters added in `fscanpftpar.c`.
- Wirth et al. (2024)  article added to `REFERENCES`.
- New parameter `"tscal_b"` added in `lpjparam.cjson`.
- New virtual function `getb()` added to type `Pft`.

### Changed

- 2 boolean flags `phen_to_one` and `rainyseason` added to tree parameter. These are used in `phenology_gsi.c` instead of string compare with tree names.
- Unused argument `aet_layer` removed from `nitrogen_stress()` function.
- `phi_bnf` parameter calculated from `swc_bnf` parameter instead of reading from `pft.cjson` file in `fscanpftpar.c`.
- Hard-coded parameter in temperature dependent calculation of `b` replaced by global parameter `tscal_b`.
- Calculation of `b` in `setupannual_grid.c` replaced by call to new virtual PFT function `getb()`.

### Removed

- Parameter `phi_bnf` removed from `pft.cjson`.
- Variable `b` removed from restart file.
- Unused macro `RESTART_VERSION` removed from `header.h`.

### Fixed

- Arguments of `npp` function corrected in `daily_woodplantation.c`. The current version leads to a too strong reduction in NPP.
- `nuptake_crop.c` changed to compile with `-DDEBUG_N` option.
- `update_daily_cell.c` changed to compile with `-DDEBUG` option.
- `daily_agriculture_tree.c` changed to compile with `-DDEBUG2` option.
- Typo in error message in `bstruct_writearrayindex.c` corrected.
- Error message corrected in `bstruct_wopen.c`.
- Error number corrected in `bstruct_finish.c`.
- Argument `temp` added to `f_turnover_tree` function to compile with `-DNRECOVERY_COST` option.


## [5.10.0] - 2025-08-14

### Contributors


src/utils/cdf2bin.c
/**************************************************************************************/
/**                                                                                \n**/
/**                     c  d  f  2  b  i  n  .  c                                  \n**/
/**                                                                                \n**/
/**     Program converts NetCDF output data into raw binary format                 \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#ifdef USE_UDUNITS
#define USAGE "Usage: %s [-h] [-swap] [-v] [-units unit] [-var name] [-map name] [-clm]\n       [-cellsize s] [-byte] [-floatgrid] [-doublegrid]  [-o binfile]\n       [-json] gridfile netcdffile ...\n"
#else
#define USAGE "Usage: %s [-h] [-swap] [-v] [-var name] [-map name] [-clm] [-cellsize s] [-byte] [-floatgrid] [-doublegrid] [-o binfile] [-json] gridfile netcdffile ...\n"
#endif
#define ERR_USAGE USAGE "\nTry \"%s --help\" for more information.\n"

#ifdef USE_NETCDF
#include <netcdf.h>

static Bool readmydata(Climatefile *file,    /* climate data file */
                       FILE *bin,            /* pointer to output file */
                       const Coord coords[], /* coordinates */
                       Bool isbyte,
                       const Config *config  /* LPJ configuration */
                      )                      /* returns TRUE on error */
{
  int t,cell,rc,index,n,start;
  size_t i;
  float *f=NULL;
  short *s=NULL;
  float data;
  size_t offsets[4];
  size_t address[2];
  size_t counts[4];
  Byte bdata;
  //printf("len=%d\n",(int)file->var_len);
  if(file->time_step==MISSING_TIME)
    start=0;
  else
  {
    start=1;
    counts[0]=1;
  }
  if(file->var_len>1)
  {
    counts[start]=1;
    index=start+1;
  }
  else
    index=start;
  offsets[index]=offsets[index+1]=0;
  counts[index]=file->nlat;
  counts[index+1]=file->nlon;
  switch(file->datatype)
  {
    case LPJ_FLOAT:
      f=newvec(float,file->nlon*file->nlat);
      if(f==NULL)
      {
        printallocerr("data");
        closeclimate_netcdf(file,TRUE);
        return TRUE;
      }
      break;
    case LPJ_SHORT:
      s=newvec(short,file->nlon*file->nlat);
      if(s==NULL)
      {
        printallocerr("data");
        closeclimate_netcdf(file,TRUE);
        return TRUE;
      }
      break;
    default:
      fprintf(stderr,"Datatype %s not supported.\n",typenames[file->datatype]);
      closeclimate_netcdf(file,TRUE);
      return TRUE;
  }
  switch(file->time_step)
  {
    case DAY:
      n=NDAYYEAR;
      break;
    case MONTH:
      n=NMONTH;
      break;
    case YEAR: case MISSING_TIME:
      n=1;
      break;
    default:
      fprintf(stderr,"ERROR421: Time step of second not supported.\n");
      return TRUE;
  }
  for(t=0;t<file->nyear*n;t++)
  {
    if(file->time_step!=MISSING_TIME)
      offsets[0]=t;
    for(i=0;i<file->var_len;i++)
    {
      //printf("t=%d,i=%d\n",t,i);
      if(file->var_len>1)
        offsets[start]=i;
      if(file->datatype==LPJ_FLOAT)
      {
        //printf("%d %d %d %d\n",(int)offsets[0],(int)offsets[1],(int)offsets[2],(int)offsets[3]);
        //printf("%d %d %d %d\n",(int)counts[0],(int)counts[1],(int)counts[2],(int)counts[3]);
        if((rc=nc_get_vara_float(file->ncid,file->varid,offsets,counts,f)))
        {
          free(f);
          fprintf(stderr,"ERROR421: Cannot read float data: %s.\n",
                  nc_strerror(rc));
          closeclimate_netcdf(file,TRUE);
          return TRUE;
        }
      }
      else
      {
        if((rc=nc_get_vara_short(file->ncid,file->varid,offsets,counts,s)))
        {
          free(s);
          fprintf(stderr,"ERROR421: Cannot read short data: %s.\n",
                 nc_strerror(rc));
          closeclimate_netcdf(file,TRUE);
          return TRUE;
        }
      }
      for(cell=0;cell<config->ngridcell;cell++)
      {
        if(file->offset)
          address[0]=file->offset-(int)((coords[cell].lat-file->lat_min)/config->resolution.lat+0.5);
        else
          address[0]=(int)((coords[cell].lat-file->lat_min)/config->resolution.lat+0.5);
        address[1]=(int)((coords[cell].lon-file->lon_min)/config->resolution.lon+0.5);
        if(checkcoord(address,cell,coords+cell,file))
        {
          if(file->datatype==LPJ_FLOAT)
            free(f);
          else
            free(s);
          closeclimate_netcdf(file,TRUE);
          return TRUE;
        }
        if(file->datatype==LPJ_FLOAT)
        {
          if(ismissingvalue(f[file->nlon*address[0]+address[1]],file->missing_value.f))
          {
            fprintf(stderr,"ERROR423: Missing value for cell=%d (",cell);
            fprintcoord(stderr,coords+cell);
            fprintf(stderr,").\n");
            free(f);
            closeclimate_netcdf(file,TRUE);
            return TRUE;
          }
          else if(isnan(f[file->nlon*address[0]+address[1]]))
          {
            fprintf(stderr,"ERROR423: Invalid value for cell=%d (",cell);
            fprintcoord(stderr,coords+cell);
            fprintf(stderr,").\n");
            free(f);
            closeclimate_netcdf(file,TRUE);
            return TRUE;
          }
          if(isbyte)
          {
            if(file->slope*f[file->nlon*address[0]+address[1]]+file->intercept<0 || file->slope*f[file->nlon*address[0]+address[1]]+file->intercept>UCHAR_MAX)
            {
              fprintf(stderr,"WARNING423: value %g for cell=%d (",file->slope*f[file->nlon*address[0]+address[1]]+file->intercept,cell);
              fprintcoord(stderr,coords+cell);
              fprintf(stderr,") does not fit in byte.\n");
            }
            bdata=(Byte)(file->slope*f[file->nlon*address[0]+address[1]]+file->intercept);
            fwrite(&bdata,sizeof(Byte),1,bin);
          }
          else
          {
            data=(float)(file->slope*f[file->nlon*address[0]+address[1]]+file->intercept);
            fwrite(&data,sizeof(float),1,bin);
          }
        }
        else
        {
          if(s[file->nlon*address[0]+address[1]]==file->missing_value.s)
          {
            fprintf(stderr,"ERROR423: Missing value for cell=%d (",cell);
            fprintcoord(stderr,coords+cell);
            fprintf(stderr,").\n");
            free(s);
            closeclimate_netcdf(file,TRUE);
            return TRUE;
          }
          if(isbyte)
          {
            if(s[file->nlon*address[0]+address[1]]<0 || s[file->nlon*address[0]+address[1]]>UCHAR_MAX)
            {
              fprintf(stderr,"WARNING423: value %d for cell=%d (",s[file->nlon*address[0]+address[1]],cell);
              fprintcoord(stderr,coords+cell);
              fprintf(stderr,") does not fit in byte.\n");
            }
            bdata=(Byte)s[file->nlon*address[0]+address[1]];
            fwrite(&bdata,sizeof(Byte),1,bin);
          }
          else
            fwrite(s+file->nlon*address[0]+address[1],sizeof(short),1,bin);
        }
      }
    }
  }
  if(file->datatype==LPJ_FLOAT)
    free(f);
  else
    free(s);
  return FALSE;
} /* of 'readmydata' */

#endif

int main(int argc,char **argv)
{
#ifdef USE_NETCDF
  Coordfile coordfile;
  Filename coord_filename;
  Climatefile data;
  Config config;
  char *units,*var,*outname,*endptr,*out_json,*arglist,*long_name,*standard_name,*history,*source,*map_name,*title;
  const char *progname;
  Coord *grid;
  Intcoord intcoord;
  float fcoord[2];
  double dcoord[2];
  FILE *file;
  Map *map=NULL;
  int iarg,j,k;
  Attr *attrs=NULL;
  int n_attr=0,len;
  char name[NC_MAX_NAME];
  float cellsize_lon,cellsize_lat;
  Bool swap,verbose,isclm,isbyte,isjson;
  Header header;
  Filename grid_name;
  Type grid_type;
  isbyte=swap=verbose=isclm=isjson=FALSE;
  units=NULL;
  var=NULL;
  map_name=NULL;
  outname="out.bin"; /* default file name for output */
  grid_type=LPJ_SHORT;
  cellsize_lon=cellsize_lat=0.5;      /* default cell size */
  initconfig(&config);
  initsetting_netcdf(&config.netcdf);
  progname=strippath(argv[0]);
  for(iarg=1;iarg<argc;iarg++)
  {
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-h") || !strcmp(argv[iarg],"--help"))
      {
        printf("   cdf2bin (" __DATE__ ") Help\n"
               "   ==========================\n\n"
               "Convert NetCDF data into raw binary output data for LPJmL version %s\n",getversion());
        printf(USAGE
               "\nArguments:\n"
               "-h,--help    print this help text\n"
               "-v,--verbose print name of NetCDF files\n"
#ifdef USE_UDUNITS
               "-units u     set unit to convert from NetCDF file\n"
#endif
               "-var name    variable name in NetCDF file\n"
               "-map name    name of map in NetCDF file\n"
               "-clm         grid file is in CLM format, default is raw\n"
               "-byte        output data is converted in byte\n"
               "-floatgrid   set datatype of grid file to float, default is short\n"
               "-doublegrid  set datatype of grid file to double, default is short\n"
               "-cellsize s  cell size in grid file, default is 0.5\n"
               "-json        JSON metafile is created with suffix '.json'\n"
               "-o binfile   filename of raw data file written. Default is out.bin\n"
               "gridfile     filename of grid data file\n"
               "netcdffile   filename of NetCDF file(s) converted\n\n"
               "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n",
               progname);
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[iarg],"-swap"))
        swap=TRUE;
      else if(!strcmp(argv[iarg],"-v") || !strcmp(argv[iarg],"--verbose"))
        verbose=TRUE;
      else if(!strcmp(argv[iarg],"-clm"))
        isclm=TRUE;
      else if(!strcmp(argv[iarg],"-byte"))
        isbyte=TRUE;
      else if(!strcmp(argv[iarg],"-floatgrid"))
        grid_type=LPJ_FLOAT;
      else if(!strcmp(argv[iarg],"-doublegrid"))
        grid_type=LPJ_DOUBLE;
      else if(!strcmp(argv[iarg],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[iarg],"-var"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-var'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        var=strdup(argv[++iarg]);
        check(var);
      }
#ifdef USE_UDUNITS
      else if(!strcmp(argv[iarg],"-units"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-units'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        units=strdup(argv[++iarg]);
        check(units);
      }
#endif
      else if(!strcmp(argv[iarg],"-map"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-map'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        map_name=argv[++iarg];
      }
      else if(!strcmp(argv[iarg],"-o"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-o'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        outname=argv[++iarg];
      }
      else if(!strcmp(argv[iarg],"-cellsize"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-cellsize'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        cellsize_lon=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-cellsize'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
        if(cellsize_lon<=0)
        {
          fprintf(stderr,"Cell size=%g must be greater than zero.\n",cellsize_lon);
          return EXIT_FAILURE;
        }
        cellsize_lat=cellsize_lon;
      }
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                ERR_USAGE,argv[iarg],progname,progname);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  }
  if(argc<=iarg)
  {
    fprintf(stderr,"Grid and NetCDF file missing.\n"
            ERR_USAGE,progname,progname);
    return EXIT_FAILURE;
  }
  if(argc<iarg+2)
  {
    fprintf(stderr,"NetCDF file missing.\n"
            ERR_USAGE,progname,progname);
    return EXIT_FAILURE;
  }
  if(isclm)
  {
    coord_filename.name=argv[iarg];
    coord_filename.fmt=CLM;
    coordfile=opencoord(&coord_filename,TRUE);
    if(coordfile==NULL)
      return EXIT_FAILURE;
    grid=newvec(Coord,numcoord(coordfile));
    if(grid==NULL)
    {
      printallocerr("grid");
      return EXIT_FAILURE;
    }
    config.ngridcell=numcoord(coordfile);
    grid_type=getcoordtype(coordfile);
    if(config.ngridcell==0)
    {
      fprintf(stderr,"Number of cells is zero in '%s'.\n",argv[iarg]);
      return EXIT_FAILURE;
    }
    getcellsizecoord(&cellsize_lon,&cellsize_lat,coordfile);
    config.resolution.lat=cellsize_lat;
    config.resolution.lon=cellsize_lon;
    for(j=0;j<numcoord(coordfile);j++)
      readcoord(coordfile,grid+j,&config.resolution);
    header.firstcell=getfirstcoord(coordfile);
    closecoord(coordfile);
  }
  else
  {
    file=fopen(argv[iarg],"rb");
    if(file==NULL)
    {
      fprintf(stderr,"Error opening grid file '%s': %s.\n",argv[iarg],
              strerror(errno));
      return EXIT_FAILURE;
    }
    switch(grid_type)
    {
       case LPJ_SHORT:
         if(getfilesizep(file) % (sizeof(short)*2)!=0)
           fprintf(stderr,"Warning: File size of '%s' is not multiple of %d.\n",argv[iarg],(int)(sizeof(short)*2));
         config.ngridcell=getfilesizep(file)/sizeof(short)/2;
         break;
       case LPJ_FLOAT:
         if(getfilesizep(file) % (sizeof(float)*2)!=0)
           fprintf(stderr,"Warning: File size of '%s' is not multiple of %d.\n",argv[iarg],(int)(sizeof(float)*2));
         config.ngridcell=getfilesizep(file)/sizeof(float)/2;
         break;
       case LPJ_DOUBLE:
         if(getfilesizep(file) % (sizeof(double)*2)!=0)
           fprintf(stderr,"Warning: File size of '%s' is not multiple of %d.\n",argv[iarg],(int)(sizeof(double)*2));
         config.ngridcell=getfilesizep(file)/sizeof(double)/2;
         break;
       default:
         fprintf(stderr,"Invalid datatype %d in '%s'.\n",grid_type,argv[iarg]);
         return TRUE;
    }
    if(config.ngridcell==0)
    {
      fprintf(stderr,"Number of cells is zero in '%s'.\n",argv[iarg]);
      return EXIT_FAILURE;
    }
    grid=newvec(Coord,config.ngridcell);
    if(grid==NULL)
    {
      printallocerr("grid");
      return EXIT_FAILURE;
    }
    switch(grid_type)
    {
      case LPJ_SHORT:
        for(j=0;j<config.ngridcell;j++)
        {
          readintcoord(file,&intcoord,swap);
          grid[j].lat=intcoord.lat*0.01;
          grid[j].lon=intcoord.lon*0.01;
        }
        break;
      case LPJ_FLOAT:
        for(j=0;j<config.ngridcell;j++)
        {
          freadfloat(fcoord,2,swap,file);
          grid[j].lon=fcoord[0];
          grid[j].lat=fcoord[1];
        }
        break;
      case LPJ_DOUBLE:
        for(j=0;j<config.ngridcell;j++)
        {
          freaddouble(dcoord,2,swap,file);
          grid[j].lon=dcoord[0];
          grid[j].lat=dcoord[1];
        }
        break;
      default:
        /* do nothing */
        break;
    }
    config.resolution.lat=cellsize_lat;
    config.resolution.lon=cellsize_lon;
    header.firstcell=0;
    fclose(file);
  }
  header.cellsize_lat=(float)config.resolution.lat;
  header.cellsize_lon=(float)config.resolution.lon;
  header.ncell=config.ngridcell;
  file=fopen(outname,"wb");
  if(file==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",outname,strerror(errno));
    return EXIT_FAILURE;
  }
  data.oneyear=FALSE;
  for(j=iarg+1;j<argc;j++)
  {
    if(verbose)
      printf("%s\n",argv[j]);
    if(openclimate_netcdf(&data,argv[j],NULL,var,NULL,units,&config))
    {
      fprintf(stderr,"Error opening '%s'.\n",argv[j]);
      return EXIT_FAILURE;
    }
    if(var==NULL)
      var=getvarname_netcdf(&data);
    if(units==NULL)
      units=getattr_netcdf(data.ncid,data.varid,"units");
    long_name=getattr_netcdf(data.ncid,data.varid,"long_name");
    standard_name=getattr_netcdf(data.ncid,data.varid,"standard_name");
    history=getattr_netcdf(data.ncid,NC_GLOBAL,"history");
    source=getattr_netcdf(data.ncid,NC_GLOBAL,"source");
    title=getattr_netcdf(data.ncid,NC_GLOBAL,"title");
    if(j==iarg+1)
    {
      if(isjson)
      {
        if(map_name!=NULL)
        {
          map=readmap_netcdf(data.ncid,map_name);
          if(map==NULL)
          {
            fprintf(stderr,"Map '%s' not found in '%s'.\n",map_name,argv[j]);
            map_name=NULL;
          }
          else
            map_name=BAND_NAMES;
        }
        else if((map=readmap_netcdf(data.ncid,config.netcdf.pft_name.name))!=NULL)
          map_name=BAND_NAMES;
        else if((map=readmap_netcdf(data.ncid,config.netcdf.depth.name))!=NULL)
          map_name=BAND_NAMES;
        if(nc_inq_natts(data.ncid,&len))
          n_attr=0;
        else
        {
          attrs=newvec(Attr,len);
          if(attrs==NULL)
          {
            printallocerr("attrs");
            return EXIT_FAILURE;
          }
          n_attr=0;
          for(k=0;k<len;k++)
          {
            if(!nc_inq_attname(data.ncid,NC_GLOBAL,k,name))
            {
              if(strcmp(name,"history") && strcmp(name,"source") && strcmp(name,"title"))
              {
                attrs[n_attr].value=getattr_netcdf(data.ncid,NC_GLOBAL,name);
                if(attrs[n_attr].value!=NULL)
                  attrs[n_attr++].name=strdup(name);
              }
            }
          }
        }
      }
      else
        map_name=NULL;
    }
    if(isclm || isjson)
    {
      if(j==iarg+1)
      {
        header.firstyear=data.firstyear;
        header.nyear=data.nyear;
        header.datatype=data.datatype;
        header.nbands=data.var_len;
        header.timestep=1;
        header.order=CELLSEQ;
        header.scalar=1;
        switch(data.time_step)
        {
          case DAY:
            header.nstep=NDAYYEAR;
            break;
          case MONTH:
            header.nstep=NMONTH;
            break;
          case YEAR: case MISSING_TIME:
            header.nstep=1;
            header.timestep=data.delta_year;
            break;
          default:
            break;
        }
        if(isclm)
          fwriteheader(file,&header,LPJOUTPUT_HEADER,LPJOUTPUT_VERSION);
      }
      else
      {
        if(data.firstyear!=header.firstyear+header.nyear)
        {
          fprintf(stderr,"First year %d in '%s' is not %d.\n",data.firstyear,argv[j],header.firstyear+header.nyear);
          return EXIT_FAILURE;
        }
        if(data.var_len!=header.nbands)
        {
          fprintf(stderr,"Number of bands %d in '%s' is not %d.\n",(int)data.var_len,argv[j],header.nbands);
          return EXIT_FAILURE;
        }
        header.nyear+=data.nyear;
      }
    }
    if(readmydata(&data,file,grid,isbyte,&config))
    {
      fprintf(stderr,"Error reading '%s'.\n",argv[j]);
      return EXIT_FAILURE;
    }
    closeclimate_netcdf(&data,TRUE);
  }
  if(isclm)
  {
    rewind(file);
    fwriteheader(file,&header,LPJOUTPUT_HEADER,LPJOUTPUT_VERSION);
  }
  fclose(file);
  if(isjson)
  {
    out_json=malloc(strlen(outname)+strlen(JSON_SUFFIX)+1);
    if(out_json==NULL)
    {
      printallocerr("filename");
      return EXIT_FAILURE;
    }
    strcat(strcpy(out_json,outname),JSON_SUFFIX);
    arglist=catstrvec(argv,argc);
    file=fopen(out_json,"w");
    if(file==NULL)
    {
      printfcreateerr(out_json);
      return EXIT_FAILURE;
    }
    grid_name.name=argv[iarg];
    grid_name.fmt=(isclm) ? CLM : RAW;
    fprintjson(file,outname,title,source,history,arglist,&header,map,map_name,attrs,n_attr,var,units,standard_name,long_name,&grid_name,grid_type,(isclm) ? CLM : RAW,LPJOUTPUT_HEADER,FALSE,LPJOUTPUT_VERSION);
    freeattrs(attrs,n_attr);
    free(out_json);
    free(arglist);
    fclose(file);
  }
  free(var);
  free(units);
  free(long_name);
  free(standard_name);
  free(history);
  free(source);
  freemap(map);
  free(grid);
  return EXIT_SUCCESS;
#else
  fprintf(stderr,"ERROR401: NetCDF is not supported in this version of %s.\n",argv[0]);
  return EXIT_FAILURE;
#endif
} /* of 'main' */
